clang/lib/Sema/TreeTransform.h

```diff
@@ -14636,6 +14645,20 @@ TreeTransform<Derived>::TransformLambdaExpr(LambdaExpr *E) {
                                     /*IsInstantiation*/ true);
   SavedContext.pop();
 
+  // Parts other than the capture e.g. the lambda body might still contain a
+  // pattern that an outer fold expression would expand.
+  //
+  // We don't have a way to propagate up the ContainsUnexpandedParameterPack
+  // flag from a Stmt, so we have to revisit the lambda.
+  if (!LSICopy.ContainsUnexpandedParameterPack) {
+    llvm::SmallVector<UnexpandedParameterPack> UnexpandedPacks;
+    getSema().collectUnexpandedParameterPacksFromLambda(NewCallOperator,
+                                                        UnexpandedPacks);
+    // FIXME: Should we call Sema::DiagnoseUnexpandedParameterPacks() instead?
+    // Unfortunately, that requires the LambdaScopeInfo to exist, which has been
+    // removed by ActOnFinishFunctionBody().
+    LSICopy.ContainsUnexpandedParameterPack = !UnexpandedPacks.empty();
+  }
```

Comment:
We probably need a refactor here to teach `ActOnFinishFunctionBody()` not to pop the LSI eventually. How about leaving it as an NFC?

Reply:
I am not sure I understand why we would need to call `Sema::DiagnoseUnexpandedParameterPacks` here.

I think we should be able to diagnose all unexpanded parameter packs during the initial lambda parsing, so what gives?

Also, I am not sure what you mean by leaving it as an NFC.

Regardless, you have a copy of the LSI here, so what would stop you from pushing a new Lambda Scope and copying this LSI back into it? Something like:

```C++
  LambdaScopeInfo *LSI = getSema().PushLambdaScope();
  *LSI = LSICopy;
  Sema::FunctionScopeRAII FuncScopeCleanup(getSema());
```

This is obviously an ugly hack, but I don't think the solution is to simply teach `ActOnFinishFunctionBody` not to pop the function scope.

See this FIXME for a potentially more comprehensive fix: https://github.com/llvm/llvm-project/blob/77ac07444d32668d5826ef27c24180fb10425213/clang/lib/Sema/TreeTransform.h#L14648

Reply:
That function attracts me because we seem to 1. always call it after collecting unexpanded packs; 2. also set the `LSI->ContainsUnexpandedParameterPack` flag there. However, this would probably be unnecessary if we called it during parsing.

> Also, I am not sure what you mean by leaving it as an NFC.

I was wondering if we could add an additional flag to `ActOnFinishFunctionBody()` so that we remove the scope info on the caller's own. (e.g. we could do that at the end of `TransformLambdaExpr`.) This way, we could also avoid copying LSI around the lines above. (I had intended to call `ContainsUnexpandedParameterPack()` afterward; again, even if the call is unnecessary, we still could benefit from the refactoring because we don't need to copy LSI.)

> but I don't think the solution is to simply teach ActOnFinishFunctionBody not to pop the function scope.

Can you say more? I think the FIXME is suggesting another way to avoid copying LSI, but that doesn't necessarily mean tweaking `ActOnFinishFunctionBody` is impossible?

Reply:
Despite these, I have another concern on performance issue after some deliberation. This is actually a workaround given that we don't have any flags on a `Stmt` to tell whether it contains unexpanded expressions - what we do now leads to visiting the lambda body twice, which might cause some impact in lambda instantiation, if the body is large enough.

A compromise is, we remove this traversal in this patch and *leave it for future fixes*. (Sigh!)

Reply:
The double traversal has now been removed from the patch, so I'll mark this resolved.